### PR TITLE
Update assemble to allow alternate source of files

### PR DESCRIPTION
**Pull request by itdependsnetworks**
##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

Assemble module

##### ANSIBLE VERSION

```
ansible 2.3.0 (assemble-sort 1c9218785b) last updated 2016/12/18 20:50:59 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

```

##### SUMMARY

Allow alternate to sort in the assemble module. 
Refer to: https://github.com/ansible/ansible/issues/19442

```

```

### CODE REVIEW

**Comment by itdependsnetworks** [comment]
@abadger @bcoca 

Any chance either of you can help me out here? I don't know what is unique about these environments that it is failing on, but it seems to be fine for me and all but 3 envs.

File: lib/ansible/modules/files/assemble.py [hunk 1]
```diff
@@ -33,14 +33,22 @@
...
        Puppet calls this idea I(fragments).
 version_added: "0.5"
 options:
   src:
     description:
       - An already existing directory full of source files.
-    required: true
+    required: false
+    default: null
+    aliases: []
```
Comment (gundalow) [inline lib/ansible/modules/files/assemble.py:45]: This line can be removed, as it's the default.

File: lib/ansible/modules/files/assemble.py [hunk 2]
```diff
@@ -33,14 +33,22 @@
...
+  src_files:
+    description:
+      - A list of files sorted by user. 
+    version_added: "2.3"
+    required: false
     default: null
     aliases: []
```
Comment (gundalow) [inline lib/ansible/modules/files/assemble.py:51]: This line can be removed, as it's the default.

**Review by gundalow — CHANGES_REQUESTED** [review]
needs_revision

Bonus points for adding integration tests

In the integration tests it would be good to run the same tasks again and ensure nothing is changed on the second run.

e.g.
test assemble with src_files and reordered
assert the fragments were assembled in correct order
test assemble with src_files and reordered (idempotent check)
assert that no change took place

ditto for test assemble with src_files and reordered and no src

**Comment by abadger** [comment]
@itdependsnetworks at the meeting last week I know we talked about a sorting only feature not satisfying your use case however, I don't see that use case listed here or in the previous ticket.  Could you post that so we have more information to discuss?

**Comment by itdependsnetworks** [comment]
Honestly I would prefer to open that against a separate issue so that there is no confusion as to why or why not this PR was accepted.

I know that @ogenstad was looking for sorting feature, so perhaps he can offer some insight.

**Comment by ogenstad** [comment]
What I want to do is to create configuration for network devices using the `template` along with `assemble`. In some cases the order matters and since the assemble module sorts the source files by name this wasn't optimal. What I thought of doing was to add s sort_order option to the assemble module so the source files could be sorted by:

* name
* name_r (in reverse order)
* ctime
* ctime_r
* mtime
* mtime_r

It would have worked if it wasn't for the fact that some of the files got created during the same second, so they got the same timestamp.

Thinking about a way to solve this the thing I came up with was that if there was some kind of variable like `{{ play_number }}` which was accessible and incremented for each task in a playbook I could use that as a prefix for each file I create with the template module. As far as I'm aware this isn't possible at the moment though.

**Comment by bcoca** [comment]
@ogenstad  for your use case I would just use template includes as you will be able to reuse the generation order for the include loop

**Comment by abadger** [comment]
Hello, we've been discussing this at several meetings over the course of the last month (summary here:  https://github.com/ansible/community/issues/150#issuecomment-277026033 ) and finally took a vote on it to judge whether the Core committers felt it was something they wanted to support in ansible or not.  We had 1 vote for, 3 abstain, and 3 vote against.  I think people felt that the same task could be achieved in other, existing ways.  I'll close this PR now.  Sorry that we decided not to include this particular PR into the core code base.

**Comment by itdependsnetworks** [comment]
Ironically was going to bring up another use case. Jinja include does not include any recursive directories. Is that not a valid use case? 

@abadger

**Pull request closed without merging**